flink/src/main/java/org/apache/iceberg/flink/FlinkFilters.java

```diff
@@ -0,0 +1,266 @@
+    Optional<Expression> right = convert(args.get(1));
+    if (left.isPresent() && right.isPresent()) {
+      return Optional.of(function.apply(left.get(), right.get()));
+    }
+
+    return Optional.empty();
+  }
+
+  private static Optional<Expression> convertComparisonExpression(
+      BiFunction<String, Object, Expression> function, BiFunction<String, Object, Expression> reversedFunction,
+      CallExpression call) {
+    Tuple2<String, Object> tuple2 = parseFieldAndLiteral(call);
+    if (tuple2 != null) {
+      if (literalOnRight(call.getResolvedChildren())) {
+        return Optional.of(function.apply(tuple2.f0, tuple2.f1));
+      } else {
+        return Optional.of(reversedFunction.apply(tuple2.f0, tuple2.f1));
+      }
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  private static Optional<String> toReference(org.apache.flink.table.expressions.Expression expression) {
+    return expression instanceof FieldReferenceExpression ?
+        Optional.of(((FieldReferenceExpression) expression).getName()) :
+        Optional.empty();
+  }
+
+  private static Optional<Object> toLiteral(org.apache.flink.table.expressions.Expression expression) {
+    // Not support null literal
+    return expression instanceof ValueLiteralExpression ?
+        convertLiteral((ValueLiteralExpression) expression) :
+        Optional.empty();
+  }
+
+  private static Optional<Expression> handleNaN(BiFunction<String, Object, Expression> function,
+                                                Function<String, Expression> functionNaN,
+                                                CallExpression call) {
+    Tuple2<String, Object> tuple2 = parseFieldAndLiteral(call);
+    if (tuple2 == null) {
+      return Optional.empty();
+    }
+
+    String name = tuple2.f0;
+    Object value = tuple2.f1;
+
+    if (NaNUtil.isNaN(value)) {
+      return Optional.of(functionNaN.apply(name));
+    } else {
+      return Optional.of(function.apply(name, value));
+    }
+  }
+
+  private static Optional<Object> convertLiteral(ValueLiteralExpression expression) {
+    Optional<?> value = expression.getValueAs(expression.getOutputDataType().getLogicalType().getDefaultConversion());
+    return value.map(o -> {
+      if (o instanceof LocalDateTime) {
+        return DateTimeUtil.microsFromTimestamp((LocalDateTime) o);
+      } else if (o instanceof Instant) {
+        return DateTimeUtil.microsFromInstant((Instant) o);
+      } else if (o instanceof LocalTime) {
+        return DateTimeUtil.microsFromTime((LocalTime) o);
+      } else if (o instanceof LocalDate) {
+        return DateTimeUtil.daysFromDate((LocalDate) o);
+      }
+
+      return o;
+    });
+  }
+
+  private static boolean literalOnRight(List<ResolvedExpression> args) {
+    return args.get(0) instanceof FieldReferenceExpression && args.get(1) instanceof ValueLiteralExpression;
+  }
+
+  private static Tuple2<String, Object> parseFieldAndLiteral(CallExpression call) {
```

Comment:
I think that this class can be cleaned up by rewriting this method. I think this was intended to make reuse easier, but rewriting this allows removing several methods (`literalOnRight`, `convertComparisonExpression`, `handleNaN`, `toReference`, and `toLiteral`), gets rid of the strange method contract where `Tuple2` and `null` are used, and also ensures that `instanceof` checks are done in just one place, rather than with multiple calls to `literalOnRight`.

Here's what I rewrote it to:

```java
  private static Optional<Expression> convertFieldAndLiteral(BiFunction<String, Object, Expression> expr,
                                                             CallExpression call) {
    return convertFieldAndLiteral(expr, expr, call);
  }

  private static Optional<Expression> convertFieldAndLiteral(
      BiFunction<String, Object, Expression> convertLR, BiFunction<String, Object, Expression> convertRL,
      CallExpression call) {
    List<ResolvedExpression> args = call.getResolvedChildren();
    if (args.size() != 2) {
      return Optional.empty();
    }

    org.apache.flink.table.expressions.Expression left = args.get(0);
    org.apache.flink.table.expressions.Expression right = args.get(1);

    if (left instanceof FieldReferenceExpression && right instanceof ValueLiteralExpression) {
      String name = ((FieldReferenceExpression) left).getName();
      Object lit = convertLiteral((ValueLiteralExpression) right);
      return Optional.of(convertLR.apply(name, lit));

    } else if (left instanceof ValueLiteralExpression && right instanceof FieldReferenceExpression) {
      Object lit = convertLiteral((ValueLiteralExpression) left);
      String name = ((FieldReferenceExpression) right).getName();
      return Optional.of(convertRL.apply(name, lit));
    }

    return Optional.empty();
  }
```

The comparison cases in the `switch` statement can then call this directly:

```java
      switch (op) {
        case LT:
          return convertFieldAndLiteral(Expressions::lessThan, Expressions::greaterThan, call);

        case LT_EQ:
          return convertFieldAndLiteral(Expressions::lessThanOrEqual, Expressions::greaterThanOrEqual, call);

        case GT:
          return convertFieldAndLiteral(Expressions::greaterThan, Expressions::lessThan, call);

        case GT_EQ:
          return convertFieldAndLiteral(Expressions::greaterThanOrEqual, Expressions::lessThanOrEqual, call);

        case EQ:
          return convertFieldAndLiteral((ref, lit) -> {
            if (NaNUtil.isNaN(lit)) {
              return Expressions.isNaN(ref);
            } else {
              return Expressions.equal(ref, lit);
            }
          }, call);

        case NOT_EQ:
          return convertFieldAndLiteral((ref, lit) -> {
            if (NaNUtil.isNaN(lit)) {
              return Expressions.notNaN(ref);
            } else {
              return Expressions.notEqual(ref, lit);
            }
          }, call);
```

The only problem is that `convertLike` can't call this. But that's okay because I don't think the reversed form of `LIKE` is actually supported anyway (`"a%" LIKE col` is not allowed, right?).

So I also rewrote `convertLike`:
```java
  private static Optional<Expression> convertLike(CallExpression call) {
    List<ResolvedExpression> args = call.getResolvedChildren();
    if (args.size() != 2) {
      return Optional.empty();
    }

    org.apache.flink.table.expressions.Expression left = args.get(0);
    org.apache.flink.table.expressions.Expression right = args.get(1);

    if (left instanceof FieldReferenceExpression && right instanceof ValueLiteralExpression) {
      String name = ((FieldReferenceExpression) left).getName();
      return convertLiteral((ValueLiteralExpression) right).flatMap(lit -> {
        if (lit instanceof String) {
          String pattern = (String) lit;
          Matcher matcher = STARTS_WITH_PATTERN.matcher(pattern);
          // exclude special char of LIKE
          // '_' is the wildcard of the SQL LIKE
          if (!pattern.contains("_") && matcher.matches()) {
            return Optional.of(Expressions.startsWith(name, matcher.group(1)));
          }
        }

        return Optional.empty();
      });
    }

    return Optional.empty();
  }
```

Reply:
@zhangjun0x01, I've posted my rewrite of this class here so you don't have to apply all the changes: https://gist.github.com/rdblue/486bd03df873a2094305025920f94498